File: src/main/java/com/donrobo/fpbg/planner/ProductionStep.java
package com.donrobo.fpbg.planner;

import com.donrobo.fpbg.data.Recipe;
import lombok.AllArgsConstructor;
import lombok.Data;

import java.util.List;
import java.util.stream.Collectors;

@Data
@AllArgsConstructor
public class ProductionStep {

    private final Recipe recipe;
    private final AssemblyMachine assemblyMachine;
    private final int count;

    private double getCraftingSpeed() {
        return assemblyMachine.getCraftingSpeed() * count;
    }

    public List<FractionalItemStack> getResultPerSecond() {
        return recipe.getResult().stream().map(is -> new FractionalItemStack((is.getCount() * getCraftingSpeed()) / recipe.getEnergyRequired(), is.getItem())).collect(Collectors.toList());
    }

    public List<FractionalItemStack> getIngredientsPerSecond() {
        return recipe.getIngredients().stream().map(is -> new FractionalItemStack((is.getCount() * getCraftingSpeed()) / recipe.getEnergyRequired(), is.getItem())).collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/donrobo/fpbg/planner/ProductionStep.kt
package com.donrobo.fpbg.planner

import com.donrobo.fpbg.data.Recipe

data class ProductionStep(val recipe: Recipe, val craftingSpeed: Double) {

    val resultPerSecond = FractionalItemStack(recipe.result.count * craftingSpeed / recipe.energyRequired, recipe.result.item)

    val ingredientsPerSecond = recipe.ingredients.map { FractionalItemStack(it.count * craftingSpeed / recipe.energyRequired, it.item) }

    fun requires(input: String) = recipe.ingredients.map { it.item.name }.contains(input)
}
